### ISSUE TITLE
Crash after calling logout

### ISSUE THREAD

**Issue by moubry** [GetStream/stream-chat-swift#2381]
## What did you do?

I implemented user sign out in my app. As part of that process, I'm calling `ChatClient.shared.logout()` function, so that the user session with Stream is destroyed, so if they sign in again with a different user, there’s nothing leftover from the previous Stream user session.

## What did you expect to happen?

I expected it to not crash.

## What happened instead?

Stream crashes intermittently.

## GetStream Environment
**GetStream Chat version:** 4.24.0
**GetStream Chat frameworks:** StreamChat, StreamChatUI
**iOS version:** latest
**Swift version:** latest
**Xcode version:** latest
**Device:** latest

## Additional context

I realize an intermittent crash during logout is likely not a very high priority issue. I’m only logging it on the off-chance that this is helpful information.

I cannot reliably reproduce it. However, this specific crash occurred while the StreamChatSwiftUI channel was visible, and we detected an expired user session in our app, so we destroyed the view, and called logout on the Stream client at the exact same time. I’m guessing it’s some kind of race condition.

<img width="1694" alt="Screen Shot 2022-11-19 at 9 33 10 PM" src="https://user-images.githubusercontent.com/144283/202882604-1b0a8cb5-26d8-4471-a4c6-11475b1cd756.png">

**Comment by polqf** [comment]
Hi @moubry ,

Thank you for your report. This is something that we are actually investigating and we expect to have a fix for it soon.

Long story short, log out does not finish all the "clean up" synchronously, and we discovered a potential chance of race conditions if you log in just after calling it.

**Comment by vjard33** [comment]
Same for me... logout() doesn't clean properly on 4.22.0 (same for 4.24.0)
TokenProvider seems to be broken on 4.24.0, sometimes I got some double completionHandler call from one call, really weird... That's why I'm stuck on 4.22.0 (and for these crashes...)

**Comment by vjard33** [comment]
> Hi @moubry ,
> 
> Thank you for your report. This is something that we are actually investigating and we expect to have a fix for it soon.
> 
> Long story short, log out does not finish all the "clean up" synchronously, and we discovered a potential chance of race conditions if you log in just after calling it.

@polqf @nuno-vieira 
Can you tell me if this logout() bug exists on 4.22.0? I want to upgrade to 4.24.1, but don't know if this is a good idea. I didn't see any crash on logout() on 4.22.0.

**Comment by polqf** [comment]
@vjard33 This issue will be fixed once this PR is merged: https://github.com/GetStream/stream-chat-swift/pull/2386

The root cause is that `logOut` has some async operations that were not taken into consideration before. So it can happen if you immediately call `connect` after calling `logOut`. If you were not facing crashes before, there is no reason to believe this should be the case for you now.

**Comment by virbedi** [comment]
Hello, 

We are also facing a related issue when trying to switch between our staging and production environment (each having their own stream user). 

We call `sharedClient.logout()` and try setting up the sharedClient with a new user ID and token but get this error:
```
Error ConnectionNotSuccessful in /Users/vir/Desktop/via_ios/Pods/StreamChat/Sources/StreamChat/Workers/ChatClientUpdater.swift:137 -> Connection to the API has failed.
You can read more about making a successful connection in our docs:
https://getstream.io/chat/docs/sdk/ios/uikit/getting-started/
API Error: Optional("Error WebSocket in /Users/vir/Desktop/via_ios/Pods/StreamChat/Sources/StreamChat/WebSocketClient/WebSocketClient.swift:219 -> WebSocketEngineError(reason: \"\\0\\u{05}\", code: 1005, engineError: nil)")
2022-12-12 19:55:32.890 [ERROR] [io.getStream.chat.core.web_socket_engine_queue] [AuthenticationRepository.swift:171]
```

I wanted to confirm if this error is related to whats stated in this thread?

**Comment by polqf** [comment]
@virbedi looks like there was an issue in our backend some days ago. Could you please try again. Seems like a problem that is unrelated to the one we are tackling in this issue.

**Comment by virbedi** [comment]
@polqf Thanks, I'll update my comment.

**Comment by polqf** [comment]
Hi! This has been fixed in our latest release [4.25.0](https://github.com/GetStream/stream-chat-swift/releases/tag/4.25.0)
Closing this issue for now, feel free to reopen it if you encounter any issues.
